### DISCUSSION TITLE
Implement hotkeys for cheatcodes

### DISCUSSION THREAD

**Discussion by shprotru** [ihhub/fheroes2#7635] [General]
If will support of cheat for map reveal it could help to view AI behaviour in debug reasons.

**Comment by ihhub** [comment]
We already have such mode for debugging.

**Comment by shprotru** [comment]
how to enable it?

**Comment by ihhub** [comment]
You can enable it through debug option in the configuration file. I don't remember which value exactly.

**Comment by shprotru** [comment]
Anyway, need cheatcode 32167 on black dragons, for weak players like me

**Comment by ihhub** [comment]
There's a debug mode for this as well. We won't support cheat codes in any near future.

**Comment by aanokhi** [comment]
Hi @ihhub,

Is there a specific reason to do not support cheat codes, or it is just because nobody implemented it yet? Cheat codes allow any kid to open the map, or just win (skip) scenario, or add dragons any time not having to enable (and know how to enable) a debug mode in advance. It also allows old school gamers have fun as in the original H2. I would try to implement if nobody minds.

**Comment by ihhub** [comment]
Hi @aanokhi , there was and still is not need to do this.

The implementation must support all platforms such as PS VIta or Switch which do not have keyboard on Adventure Map so the original game's way will not work. Also save format must be modified to include cheat state and none of cheaters should appear on the high score board. Moreover, cheats must not work in multiplayer mode.

**Comment by aanokhi** [comment]
@ihhub @oleg-derevenetz 
I believe there is a need. A kid or old school player may love to play with cheats and have fun. It was a part of the original game, more funny and important than than the high score board. The only real problem is keyboardless platforms, however the cheats can be implemented for windows only under an "experimental feature" flag. Can be even disabled by default.

**Comment by ihhub** [comment]
Experimental feature dialog will be removed prior 1.0. Please check [this page](https://github.com/ihhub/fheroes2/wiki/Adding-a-new-option-into-the-game).

You can check how save file dialog works for all systems.

**Comment by ihhub** [comment]
Implementing a feature only for a particular platform is not a proper way to go. We have a decent amount of Linux and MacOS players.

Like I mentioned previously we can't implement in the same way as it was in the original game. We need to find a user friendly way which will work not only on Windows but even on Android without buttons.

**Comment by ihhub** [comment]
Hi @aanokhi , I believe that we need to add a hotkey combination to open a special dialog where a player can type something. This dialog can be used in the future not only for cheats but for in-game chat as well.

For the first iteration this dialog can be somehow similar to file save dialog which supports all platforms. Upon a successful typing of a cheat we need to display a message (something like "You hear some distant noises coming from the sky. Moments later you spot silhouettes above the clouds. Something huge and terrifying is coming towards you... A group of dragons descended to you offering to join your army and you accept this offer."). Then upon ending a scenario we need to show a message to a player being a cheater and that they are not allowed to be present in High Score board.

I assume that @Branikolog  and @LeHerosInconnu can come up with even better descriptive solution for the popup dialogs.

This is how I envision the cheating system to work. Please correct me if something is wrong or you have a better solution.

**Comment by Branikolog** [comment]
Hi, @ihhub 
We definitely need a nice-looking dialog window, where we can type proper combination (with visual representation of what we're typing) and where the result of our input will be described. Also here should be a warning for players, that when they use cheat, the map result will be marked with "Cheater" label.
I personally prefer results look more formal. Something like: "The map was revealed", or "Mighty creatures joined your army".

>Moreover, cheats must not work in multiplayer mode.

Is a mandatory condition.

**Comment by ihhub** [comment]
I am moving this topic to the discussion section. This needs a huge rework due to presence of configurable hotkeys in the engine.

**Comment by Greg956** [comment]
So, what about debug mode, how to enable it?

**Comment by itreers** [comment]
Why such option presented in VCMI (with they own codes), but it's not possible to implement is here?

**Comment by oleg-derevenetz** [comment]
> Why such option presented in VCMI (with they own codes), but it's not possible to implement is here?

You are free to implement it here and propose a corresponding pull request provided that:

1. This proposal should work on all supported platforms - even on those that do not have keyboard at all (Android, game consoles such as PlayStation Vita or Nintendo Switch, etc);
2. This proposal should not conflict with the existing fheroes2 hotkey mechanism;
3. The UI of this proposal should look harmoniously as part of the HoMM2 UI.

**Comment by jabcreations** [comment]
>     1. This proposal should work on all supported platforms - even on those that do not have keyboard at all (Android, game consoles such as PlayStation Vita or Nintendo Switch, etc);

...you can't be serious. Heroes II is a PC game! You're willing to screw over legitimate players for a platform family forget about the fact that the specific device would not exist for almost two decades?!

Type a cheat code, get the result. Don't screw over literally 99.% of people because of extreme absurdity.

**Reply by oleg-derevenetz** [reply]
> Heroes II is a PC game!

Original Heroes II was a PC game, fheroes2 is a multiplatform game engine. Users of other platforms will also want to enter cheat codes.

> Don't screw over literally 99.% of people because of extreme absurdity.

Where are you getting all these numbers, Potter? (C)

**Comment by Mr-Bajs** [comment]
> > ```
> > 1. This proposal should work on all supported platforms - even on those that do not have keyboard at all (Android, game consoles such as PlayStation Vita or Nintendo Switch, etc);
> > ```
> 
> ...you can't be serious. Heroes II is a PC game! You're willing to screw over legitimate players for a platform family forget about the fact that the specific device would not exist for almost two decades?!
> 
> Type a cheat code, get the result. Don't screw over literally 99.% of people because of extreme absurdity.

I would say that Heroes of Might and Magic 2 are foremost a DOS and a Windows 95 and 98 game. Though a major point of source ports and such is game preservation and making the games playable without out of date operatingsystems and hardware.

I have a hard time seeing how a player of a free and open source game is being screwed over? There isn't anything stopping anyone from creating a fork of fheroes2 to a adding features that they want. Or just play the og game on windows 95 which where the intended platform, what is stopping them?

**Comment by Msigou01** [comment]
Is there any update on getting cheats working? Still trying to use cheats in this as the difficulty is much too hard for some levels, and it does make the game unplayable. This was part of the enjoyment in the old game.

**Reply by ihhub** [reply]
Hi @Msigou01 , are you using the latest version of the engine? We greatly reduced AI capabilities in the latest release.

**Reply by jabcreations** [reply]
You guys don't get it: the AI is fine for what it is/was. The whole point of using cheat codes was because to smooth out the rougher parts of games. We don't want people to redesign the game, only restore it.

**Reply by ihhub** [reply]
We totally get it. This is why we are easing AI for Easy difficulty so any player can complete the game. Instant battles already give a huge time reduction spending on a map. You can literally complete a map within minutes using these 2 features. After all many players play the game for the sake of playing, not to complete it asap.

I am suggesting to look at the "problem" from a different angle: what exactly you want to skip and why? Is your gameplay level is not enough? Would it help to have an in-game tutorial and hints? Be open in ideas and suggestions and we can find a compromise. All issues highlighted before for implementing cheat codes are still present.

**Reply by jabcreations** [reply]
Let me translate what you just typed:
"Would you like anything other than cheat codes?"

No, we just want the cheat codes please. You know, the last original part of the game still missing in this version. If this were a website I could implement it in less than an hour though I don't do this kind of programming. I see no valid reason that is keeping the people working on the project from implementing it. Other platforms that didn't even exist don't count and aren't a valid reason to reduce the functionality of the game on it's intended platform: PC. It's not my project though I do appreciate the efforts that other people make. You folks want to add some enhancements? Great, have at it! But the last thing anyone working on software should ever do is dictate to people any kind of how and why they should use the software. I don't want to complete a map in 2 minutes, I just don't want to complete it in 20 days because I have to wade through several days of tutorials because of an arbitrary play style of the original developers on a campaign map.

**Reply by ihhub** [reply]
Please provide statistics regarding the usage per platform as well as UI mockup to cover all of them. This project is for all platforms and many users (based on our statistics) prefer to play it on mobile phones rather than on PC.

Cheat code are the least important thing to do at the moment. There are many other issues which even break the game like passability rules or absence of the Editor and network multiplayer. Spending time for implementing cheat code and testing them for all platforms is not a right task to do at the moment.

It would be greatly appreciated if anyone can help with the development of more important features right now so that the team will have time to come to less critical tasks sooner.

We do not dictate anyone and it is fair to say that also nobody should dictate the team on what to do and how. If the intention to just play the original game on a PC (Windows only) with no chances in the gameplay then GoG version is a perfect fit for this.

**Reply by shprotru** [reply]
Hi @jabcreations ! I am a weak player, and I would like to see the opportunity to play with cheat codes. I've been thinking a lot about cheats and I've always been tormented by the following point: in the project roadmap is implementing a network game. How do you assess: will the ability to enter cheats undermine confidence in the integrity of the online game, or not? It seems to me that this is a double-edged sword. It's better to have a good game without cheats, than a game without playing through the network but with cheats.

**Reply by Msigou01** [reply]
Pretty much not including the cheats is a deviation from the original game and therefore the intention of why some people are wanting to play this game. So I do view it as a critical issue.

**Comment by jabcreations** [comment]
As a developer it is a simple as checking a status as on or off, single or multiplayer. Programming not setup that way? Then iterate through the players to determine if two or more are not AI players.

Secondly other platforms are not valid. This is akin to insisting that the color purple has a smell. It's far beyond the scope of the game.

The point of the cheat codes is the nullify the shortcomings of the game and to still be able to enjoy it within a reasonable amount of time. Since this exchange is proving pointless through the sheer amount of time being wasted by pointless obstinance and every lame excuse that could possibly be conjured I'm forced to consider this project dead.

**Reply by ihhub** [reply]
As I explained earlier we appreciate help in development. Please come up with a UI mockup to support all platforms as well a well-written code for a pull request which includes implementation of all cheat codes, save file corrections, migration from old save formats to new ones and UI for High Scores. I pointed out that PC is not the only platform to be used. This is based on facts, not on assumptions.

**Comment by Msigou01** [comment]
Can anyone help get this working so we can enable cheats? As someone who can’t really edit games much and is a normal user, I’d love to see these changes implemented, even as quick solution for Mac/Pc Users

**Comment by sroziewski** [comment]
How can I enable the debug mode, any examples?

**Reply by sroziewski** [reply]
I have to tried to enable debug mode, but without any result.
I have complied the project with `cmake -B build && cmake --build build --config debug --clean-first --parallel 10`
and changed the debug code in a cfg file, with no effect.
However, I have come up with an idea of changing artifact traits given that I knew that on a map there is a certain artifact nearby.
Then compile and load the saved game.

I have put in the file `fheroes2-1.0.13/src/fheroes2/resource/artifact_info.cpp`

` artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::MORALE, 10 );
        artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::LUCK, 10 );
        artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::GOLD_INCOME, 10000 );
        artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::SPELL_POWER_SKILL, 1000 );
        artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::DEFENCE_SKILL, 1000 );
        artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::ATTACK_SKILL, 1000 );
        artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::KNOWLEDGE_SKILL, 1000 );`

        artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::EXTRA_CATAPULT_SHOTS, 5 );
        artifactData[Artifact::MEDAL_HONOR].bonuses.emplace_back( fheroes2::ArtifactBonusType::LAND_MOBILITY, 300 );

**Reply by ihhub** [reply]
@sroziewski , don't you think that AI will **also** have these advantages after picking an artifact? AI knows the real value of each artifact even if you change its stats.

**Reply by sroziewski** [reply]
![image](https://github.com/ihhub/fheroes2/assets/942525/5cfa5e8b-cce6-4ca3-9cd4-f1e9b286f1ff)

**Reply by oleg-derevenetz** [reply]
> and changed the debug code in a cfg file, with no effect.

How exactly did you determine that there was no effect? Debug mode is mainly designed for debugging, i.e. there will be more verbose logging (what exactly will be logged depends on on the exact value of the `debug` parameter), and, at some values (when the `DBG_DEVEL` bit is set), the entire map will also be revealed and Debug Hero will be "hired" at the start of the map by the first human player in a row.

**Comment by sroziewski** [comment]
> @sroziewski , don't you think that AI will **also** have these advantages after picking an artifact? AI knows the real value of each artifact even if you change its stats.

I didn't know that fact, I have just tried to find some way to interact with the game state.
This artifact was really close to my castle so it was quite possible that I would be the first guy to reach it.

**Reply by ihhub** [reply]
Modding is one of the ways to break the game. You can modify the code but you must be sure about what you are doing and what outcomes your changes lead to :)

**Reply by sroziewski** [reply]
I have had the chance to look deeper into code, that's quite interesting.
